netcore/System.Private.CoreLib/src/System/TypeNameParser.cs

```diff
@@ -157,25 +157,40 @@ static Type ResolveType (Assembly assembly, List<string> names, Func<Assembly, s
 						throw new TypeLoadException (SR.Format (SR.TypeLoad_ResolveNestedType, names[i], names[i-1]));
 					else
 						break;
-                }
+				}
 			}
 			return type;
 		}
 
-        static string EscapeTypeName (string name)
-        {
-            if (name.IndexOfAny (SPECIAL_CHARS) < 0)
-                return name;
+		static string EscapeTypeName (string name)
+		{
+			if (name.IndexOfAny (SPECIAL_CHARS) < 0)
+				return name;
+
+			var sb = new StringBuilder ();
+			foreach (char c in name) {
+				if (Array.IndexOf<char> (SPECIAL_CHARS, c) >= 0)
+					sb.Append ('\\');
+				sb.Append (c);
+			}
+
+			return sb.ToString ();
+		}
+
+		static string UnescapeTypeName (string name)
+		{
+			if (name.IndexOfAny (SPECIAL_CHARS) < 0)
+				return name;
 
 			var sb = new StringBuilder ();
```

Comment:
nit: can give the string builder an initial capacity here so that it doesn't have to reallocate too often.